I2S driver for STM32F4

ledneczki opened the pull request:
Setting up I2S driver is still in a very early stage. I go slowly, step-by-step to get to know the structure of modm.

ledneczki committed:
[I2S] Setup PLL for the I2S peripheral

Functions are added to be able
to set up the I2S PLL at startup.
PLL multipliers are taken over from ST's example:
STM32F4-Discovery_FW_V1.1.0/Project/Audio_playback_and_record/src/system_stm32f4xx.c

ledneczki committed:
[I2S] Setup basic I2S driver template package

rleh commented on src/modm/platform/clock/stm32/rcc.cpp.in:
```diff
@@ -290,6 +290,29 @@ modm::platform::Rcc::enablePll(PllSource source, const PllFactors& pllFactors, u
 
+%% if target["family"] == "f4"
+bool
+modm::platform::Rcc::enablePllI2s(PllSource, const PllI2sFactors& pllFactors, uint32_t waitCycles)
```
`PllSource` is not used (because it is not configurable?).

ledneczki replied:
Good observation! Initially, I thought I will use it to do some sanity check on the frequency of the input clock of the I2S PLL. But for now the user has to check manually that the right input clock is provided. The parameter is removed.

ledneczki committed:
[I2S] Add pin configuration for I2S DAC

Remove not used parameter from I2S PLL initializer function.

Pin setup for the I2S DAC can be done
by calling Board::initializeCs43()

ledneczki committed:
[I2S] Add first draft of I2S master configuration

Clock divider for the I2S peripheral is done compile time.

ledneczki committed:
[I2S] Add audio DAC driver and interrupt driven I2S

I2S master driver can be used with the TXE (transmit buffer empty)
to transmit audio samples.

The DAC driver is not that flexible at this moment. It just provides
enough config that it can output audio on the headphone output.
Register field definition is also not complete. I just added that
much that is currently
enough to make it work.

salkinium commented on src/modm/driver/dac/cs43l22_impl.hpp:
```diff
@@ -0,0 +1,97 @@
+{
+	RF_BEGIN();
+	// Verify the chip ID
+	success = RF_CALL_BLOCKING(readRegister(Register::ChipIdRevision));
```
Just `RF_CALL` otherwise it's a synchronous blocking call.

ledneczki replied:
I replaced all the blocking calls. It took me a bit to get my head around resumable functions.

ledneczki committed:
[I2S] Fix audio DAC driver

Several fixes have been applied to
the driver, like removal of blocking calls.

ANother fix is applied to the volume setting.
The volume can be also set by specifying relative steps.

salkinium commented:
@ledneczki, what's the status of this PR? I'm seeing some good additions to the PLL code, a driver for the CS43L22 with at least the I2C interface looking somewhat complete, and parts (?) of an I2S driver? Do you want a review?

Is there something you want to merge separately already, even if it's incomplete? That way others can take over your work in the future.

ledneczki commented:
@salkinium thank you for your question! First of all excuse me for being under water for so long. I actually created an example project with stm32f4 discovery kit, which acts as a drum kit, and also the mic signal is piped into the audio DAC. I had/still have troubles finding audio samples with proper licensing (this made me a bit discouraged), so I can not push the latest code yet. But I think I am going to replace the drum kit audio samples with simple sine waves. I will try to make that happen soon.

salkinium commented:
> also the mic signal is piped into the audio DAC

Ohhhh, fancy! Looking forward to that.

> I think I am going to replace the drum kit audio samples with simple sine waves.

Yes, I think that's perfectly fine for an example. Not that there's already a [SIN table in the DAC examples](https://github.com/modm-io/modm/blob/develop/examples/nucleo_f446ze/dac_basic/main.cpp#L19-L30).

ledneczki commented:
Thanks @salkinium for the link!

Does someone know by chance if it is possible to check in third-party lib, like ST's PDM filter library (already in compiled format)?

salkinium commented:
> Does someone know by chance if it is possible to check in third-party lib, like ST's PDM filter library (already in compiled format)?

It's [technically possible via lbuild](https://github.com/modm-io/Invensense-eMD/blob/master/eMD/mpl/module.lb#L32-L35), but perhaps it should be a separate application repository. The example should just showcase the basic functionality as a starting point, what you do then with the data is application specific.

ledneczki commented:
I ask this, because the example makes use of the filters in the library in order to retrieve the audio signal from the PDM signal that is coming from the mic, but I do not know how much people in modm repo want to see a compiled third-party lib checked in.

salkinium commented:
Are you talking about the [PDM2PCM ST Audio_Addon library](https://github.com/STMicroelectronics/STM32CubeF4/tree/2f3b26f16559f7af495727a98253067a31182cfc/Middlewares/ST/STM32_Audio/Addons/PDM)? It seems to have some weird license with registering, so I'm not even sure if we allowed to host it.

ledneczki commented:
> Are you talking about the [PDM2PCM ST Audio_Addon library](https://github.com/STMicroelectronics/STM32CubeF4/tree/2f3b26f16559f7af495727a98253067a31182cfc/Middlewares/ST/STM32_Audio/Addons/PDM)? It seems to have some weird license with registering, so I'm not even sure if we allowed to host it.

Yes, indeed. Then I will try to do this conversion with the available filters.
